spi shell: macro RUN_FN_ON_SPI_DEVICE scope is wrong, e.g. for MFD devices

### Describe the bug

The problem is that the `spi_shell.c` macro `RUN_FN_ON_SPI_DEVICE` is incorrectly including SPI child nodes because it only checks if a node is "on the SPI bus" (`DT_ON_BUS(node_id, spi)`), which is true for children of SPI devices.

The issue came in with Zephyr v4.3.0. It was no issue in v4.2.1.

### Regression

- [ ] This is a regression.

### Steps to reproduce

Add Multi Function Device including GPIO controller sub-device to device tree:
```
&spi1 {
	pinctrl-0 = <&spi1_nss_pa4 &spi1_sck_pa1
		     &spi1_miso_pa6 &spi1_mosi_pe15>;
	pinctrl-names = "default";
	status = "okay";

	ad559x: ad559x@0 {
		compatible = "adi,ad559x";
		reg = <0>;
		reset-gpios = <&gpiob 9 GPIO_ACTIVE_LOW>;
		spi-max-frequency = <2000000>;
		status = "okay";

		ad559x_gpios: gpio-controller {
			compatible = "adi,ad559x-gpio";
			gpio-controller;
			#gpio-cells = <2>;
			status = "okay";
		};
	};
};
```
and to prj.conf:
```
CONFIG_MFD=y
CONFIG_MFD_AD559X=y
CONFIG_GPIO_AD559X=y
```
Enable SPI Shell:
```
CONFIG_SPI=y
CONFIG_SPI_SHELL=y
```

### Relevant log output

```shell
In file included from /workdir/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /workdir/zephyr/include/zephyr/irq_multilevel.h:16,
                 from /workdir/zephyr/include/zephyr/devicetree.h:20,
                 from /workdir/zephyr/include/zephyr/device.h:12,
                 from /workdir/zephyr/include/zephyr/drivers/spi.h:28,
                 from /workdir/zephyr/drivers/spi/spi_shell.c:12:
/workdir/myapp/build/zephyr/include/generated/zephyr/devicetree_generated.h:26066:3126: error: 'DT_N_S_soc_S_spi_40013000_S_ad559x_0_S_gpio_controller_P_spi_max_frequency' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_spi_40013000_S_ad559x_0_S_gpio_controller_P_gpio_controller'?
```

### Impact

Functional Limitation – Some features not working as expected, but system usable.

### Environment

- Zephyr SDK 0.17.4
- Zephyr Version 4.3.0
- tested with STM32U545

### Additional Context

_No response_

--- a/drivers/spi/spi_shell.c
+++ b/drivers/spi/spi_shell.c
@@ -29,11 +29,16 @@
 /* Maximum bytes we can write and read at once */
 #define MAX_SPI_BYTES MIN((CONFIG_SHELL_ARGC_MAX - TXRX_ARGV_BYTES), 32)
 
-/* Runs the given fn only if the node_id belongs to a spi device, which is on an okay spi bus. */
-#define RUN_FN_ON_SPI_DEVICE(node_id, fn)                                                          \
-	COND_CODE_1(DT_ON_BUS(node_id, spi), \
-	(COND_CODE_1(DT_NODE_HAS_STATUS_OKAY(DT_BUS(node_id)), \
-	(fn), ())), ())
+/* Runs the given fn only if the node_id is a direct child of the SPI
+ * controller (i.e. DT_PARENT(node_id) == DT_BUS(node_id)), and the node
+ * itself has status okay. This prevents enumerating sub-nodes of SPI
+ * devices (for example GPIO sub-devices) which are on the same SPI bus
+ * but are not direct children of the controller.
+ */
+#define RUN_FN_ON_SPI_DEVICE(node_id, fn)                                                         \
+	COND_CODE_1(DT_ON_BUS(node_id, spi),                                                      \
+	(COND_CODE_1(DT_SAME_NODE(DT_PARENT(node_id), DT_BUS(node_id)),                           \
+	(COND_CODE_1(DT_NODE_HAS_STATUS_OKAY(node_id), (fn), ())), ())), ())
 
 /* Create specified number of empty structs, separated by ',' */
 #define _EMPTY_STRUCT_INST(idx, list) {0}

